tasks/imagemin.js

```diff
@@ -1,82 +1,83 @@
+module.exports = grunt => {
 	grunt.registerMultiTask('imagemin', 'Minify PNG, JPEG, GIF and SVG images', function () {
-		var done = this.async();
-		var files = this.files;
-		var totalSaved = 0;
-		var options = this.options({
+		const done = this.async();
+		const options = this.options({
 			interlaced: true,
 			optimizationLevel: 3,
 			progressive: true
 		});
 
-		async.eachLimit(files, os.cpus().length, function (file, next) {
-			var msg;
-			var imagemin = new Imagemin()
-				.src(file.src[0])
-				.dest(path.dirname(file.dest))
-				.use(Imagemin.jpegtran(options))
-				.use(Imagemin.gifsicle(options))
-				.use(Imagemin.optipng(options))
-				.use(Imagemin.svgo({plugins: options.svgoPlugins || []}));
-
-			if (options.use) {
-				options.use.forEach(imagemin.use.bind(imagemin));
-			}
-
-			if (path.basename(file.src[0]) !== path.basename(file.dest)) {
-				imagemin.use(rename(path.basename(file.dest)));
-			}
+		if (Array.isArray(options.svgoPlugins)) {
+			options.plugins = options.svgoPlugins;
+		}
 
-			fs.stat(file.src[0], function (err, stats) {
-				if (err) {
-					grunt.warn(err + ' in file ' + file.src[0]);
-					return next();
-				}
+		const plugins = options.use || getDefaultPlugins(grunt, options);
 
-				imagemin.run(function (err, data) {
-					if (err) {
-						grunt.warn(err + ' in file ' + file.src[0]);
-						return next();
-					}
+		let totalBytes = 0;
+		let totalSavedBytes = 0;
+		let totalFiles = 0;
 
-					var origSize = stats.size;
-					var diffSize = origSize - ((data[0].contents && data[0].contents.length) || 0);
+		pMap(this.files, file => Promise.resolve(grunt.file.read(file.src[0], {encoding: null}))
+			.then(buf => Promise.all([imagemin.buffer(buf, {plugins}), buf]))
+			.then(res => {
+				const optimizedBuf = res[0];
+				const originalBuf = res[1];
+				const originalSize = originalBuf.length;
+				const optimizedSize = optimizedBuf.length;
+				const saved = originalSize - optimizedSize;
+				const percent = originalSize > 0 ? (saved / originalSize) * 100 : 0;
+				const savedMsg = `saved ${prettyBytes(saved)} - ${percent.toFixed(1).replace(/\.0$/, '')}%`;
+				const msg = saved > 0 ? savedMsg : 'already optimized';
 
-					totalSaved += diffSize;
-
-					if (diffSize < 10) {
-						msg = 'already optimized';
-					} else {
-						msg = [
-							'saved ' + prettyBytes(diffSize) + ' -',
-							(diffSize / origSize * 100).toFixed() + '%'
-						].join(' ');
-					}
+				if (saved > 0) {
+					totalBytes += originalSize;
+					totalSavedBytes += saved;
+					totalFiles++;
+				}
 
-					grunt.verbose.writeln(chalk.green('✔ ') + file.src[0] + chalk.gray(' (' + msg + ')'));
-					process.nextTick(next);
-				});
-			});
-		}, function (err) {
-			if (err) {
-				grunt.warn(err);
+				grunt.file.write(file.dest, optimizedBuf);
+				grunt.verbose.writeln(chalk.green('✔ ') + file.src[0] + chalk.gray(` (${msg})`));
+			})
+			.catch(err => {
+				grunt.warn(`${err} in file ${file.src[0]}`);
+			}), {
```

Comment:
The code here is a bit hard to read. Can you extract line 49 to here out into a separate variable?